src/client/Client.cc

```diff
@@ -2668,6 +2668,10 @@ void Client::handle_mds_map(MMDSMap* m)
     } else if (mdsmap->get_addrs(mds) != session->addrs) {
       session->con->mark_down();
       session->addrs = mdsmap->get_addrs(mds);
+      if (mds_sessions.count(p->first)) {
+	   ldout(cct, 1) << "handle_mds_map update connection " << dendl;
+	   session->con = messenger->connect_to_mds(session->addrs);
```

Comment:
The client really don't need reconnect to MDS before it reaches STATE_RECONNECT. But we may miss this state because network is not ok. The root cause is we only marked down Pipe when the addresses changed,the Session is still STATE_OPEN. So we need update connection so that the session can be closed in later logic.

Reply:
> The client really don't need reconnect to MDS before it reaches STATE_RECONNECT. But we may miss this state because network is not ok. 

We can detect this situation though. Try this:

````
diff --git a/src/client/Client.cc b/src/client/Client.cc
index 82577eff64f..98608349572 100644
--- a/src/client/Client.cc
+++ b/src/client/Client.cc
@@ -2663,9 +2663,7 @@ void Client::handle_mds_map(MMDSMap* m)
 
     int oldstate = oldmap->get_state(mds);
     int newstate = mdsmap->get_state(mds);
-    if (!mdsmap->is_up(mds)) {
-      session->con->mark_down();
-    } else if (mdsmap->get_addrs(mds) != session->addrs) {
+    if (!mdsmap->is_up(mds) || mdsmap->get_addrs(mds) != session->addrs) {
       session->con->mark_down();
       session->addrs = mdsmap->get_addrs(mds);
       // When new MDS starts to take over, notify kernel to trim unused entries
@@ -2674,10 +2672,21 @@ void Client::handle_mds_map(MMDSMap* m)
       trim_cache_for_reconnect(session);
     } else if (oldstate == newstate)
       continue;  // no change
-    
+
+    /* If we get here, the MDS has changed state or (!) another MDS has taken
+     * over. (Keep in mind that it's possible (oldstate == newstate ==
+     * STATE_ACTIVE) but the addr has changed! This happens when the client
+     * misses epochs.
+     */
+
     session->mds_state = newstate;
-    if (newstate == MDSMap::STATE_RECONNECT) {
+    if (newstate >= MDSMap::STATE_RECONNECT) {
       session->con = messenger->connect_to_mds(session->addrs);
+      /* If we missed reconnect, the MDS will reset the session and we'll only
+       * reconnect if client_reconnect_stale==true.
+       */
+    }
+    if (newstate == MDSMap::STATE_RECONNECT) {
       send_reconnect(session);
     } else if (newstate >= MDSMap::STATE_ACTIVE) {
       if (oldstate < MDSMap::STATE_ACTIVE) {
````